src/progress_bar.rs

```diff
@@ -522,7 +523,8 @@ impl MultiProgress {
         let (tx, rx) = channel();
         MultiProgress {
             state: Arc::new(RwLock::new(MultiProgressState {
-                objects: vec![],
+                objects: Vec::new(),
+                free_set: HashSet::new(),
```

Comment:
I think we can use a `BinaryHeap` for the free set, or do you see an actual risk of duplicates?

Reply:
I was guided by the performance of `pop`, which in case of `HashSet` is O(1), it is just
```rust
let Some(i) = state.free_set.iter().next().cloned()
```
and in case of other data structures not that fast (e.g. `BinHeap`'s `pop` operation is O(log). Fortunately we don't have to choose the minimum tombstone, just some of them if any, since the visual order is kept maintained by `ordering` array.

The duplication can be done by misuse, with e.g.:
```
let pb = mp.add(ProgressBar::new(10));
mp.remove(&pb);
mp.remove(&pb);
```
After the second `remove` the `free_set` would contain 2 zeros, if it allowed duplications.
I think I add the test to control the correct behavior

Reply:
We can prevent the duplication by checking at remove time whether something is present in the slot (`is_some()`), and only adding it to the free set if it was previously occupied.

I'm not sure the `O(log)` complexity of `BinaryHeap::pop()` should be guiding? In your project, how does the number of progress bar additions relate to the number of removals? If `pop()` is cheaper but `push()` is substantially more expensive (which I think would be the case for `HashSet` compared to `BinaryHeap`), the `HashSet` doesn't work to our advantage (even ignoring the added memory usage)

Reply:
I can imagine the number of progress bars like 100-150 in worst cases. I think `BinaryHeap` works for our case too, I will make the change.

Reply:
That's the number, but I meant are additions/removals generally 1:1? Or do you expect that there will be more like 100 additions and 25 removals?

Reply:
Let me show
<img width="746" alt="image" src="https://user-images.githubusercontent.com/940363/113738559-d5fff080-9720-11eb-9319-6419826546f6.png">
This is all complete state, but in reality many of tree nodes are working and the progress bars show their statuses.

Each collapse/expand removes/adds a bunch of progress bars at once, the extreme case is to collapse/expand the root node, the operation removes and adds almost all progress bars.
100 lines on the terminal we can do using smaller fonts, I think 150 is the upper bound on the `MultiProgress` size, in case of big trees we will implement some workarounds (scrolls or so).